keps/sig-node/4540-strict-cpu-reservation/README.md

```diff
@@ -0,0 +1,498 @@
+
+### Non-Goals
+
+## Proposal
+
+We propose to add a new `CPUManager` policy option called `strict-cpu-reservation` to the `static` policy of `CPUManager`.
+When this policy option is enabled, we remove the reserved cores from the list of all available cores at the stage of calculation DefaultCPUSet. As a result, burstable and best-effort containers are launched with a cpuset in which the reserved cores are excluded.
+
+### User Stories (Optional)
+
+#### Story 1
+To protect latency of workload, systemd daemons including irqbalance daemon are commonly constrained to the reserved CPUs.
+Burstable and best-effort pods (and guaranteed pods with fractional CPU requests) running on the reserved CPUs causes CPU throttling for infrastructure services which results in poor system response time which in turn hits back on workload response time.
+This issue is particularly bad in all-in-one deployments where workloads are placed on combined master+worker+storage nodes.
+
+#### Story 2
+Silently allowing workloads running on the reserved CPUs makes benchmarking infrastructure and workloads both inaccurate.
+
+### Risks and Mitigations
+
+The feature is isolated to a specific policy option `strict-cpu-reservation` under `cpuManagerPolicyOptions` and is protected by feature gate `CPUManagerPolicyAlphaOptions` or `CPUManagerPolicyBetaOptions` before the feature graduates to `Stable` i.e. enabled by default.
+
+Concern for feature impact on best-effort workloads, the workloads that do not have resource requests, is brought up.
+
+Kube-scheduler schedules pods on node allocatable (total - reserved). For best-effort pods, kube-scheduler uses default request values when scoring the nodes, see https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/util/pod_resources.go#L32 and https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/framework/plugins/noderesources/resource_allocation.go#L123, but the scheduler does not use the default request values when fitting the nodes i.e. best-effort pods are always admitted.
+
+The concern is, when the feature graduates to `Stable`, it will be enabled by default, best-effort workloads could be starved on the node when the node runs out of non-reserved CPU cores.
+
+However, this is exactly the feature intent, best-effort workloads have no KPI requirement, they are meant to consume whatever CPU resources left on the node including starving from time to time. Best-effort workloads are not scheduled to run on the `reservedSystemCPUs` so they shall not be run on the `reservedSystemCPUs` to destablize the whole node.
+
+Nevertheless, risk mitigation, or the need for it, is under discussion.
+
+## Design Details
+
+In Kubelet, when `strict-cpu-reservation` is enabled as a policy option, we remove the reserved cores from the shared pool at the stage of calculation DefaultCPUSet and remove the `MinSharedCPUs` from the list of available cores for exclusive allocation.
```

Comment:
This reads a bit confusing to me. Can we start that we are proposing introducing a MinSharedCPUs option? Maybe start with the default proposal and move MinSharedCPUs option under a sub section to resolve BE starvation issue.

Reply:
Sorry for confusion. The minSharedCPUs is under the original risk mitigation option and I forgot to remove it when updating the KEP.